source/GStreamerWebAudioPlayerClient.cpp

```diff
@@ -0,0 +1,339 @@
+            {
+                GST_ERROR("Could not create web audio backend");
+                mIsOpen = false;
+            }
+            result = mIsOpen;
+        });
+
+    return result;
+}
+
+bool GStreamerWebAudioPlayerClient::play()
+{
+    GST_DEBUG("entry:");
+
+    bool result = false;
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            if (mIsOpen)
+            {
+                result = mClientBackend->play();
+            }
+            else
+            {
+                GST_ERROR("No web audio backend");
+            }
+        });
+
+    return result;
+}
+
+bool GStreamerWebAudioPlayerClient::pause()
+{
+    GST_DEBUG("entry:");
+
+    bool result = false;
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            if (mIsOpen)
+            {
+                result = mClientBackend->pause();
+            }
+            else
+            {
+                GST_ERROR("No web audio backend");
+            }
+        });
+
+    return result;
+}
+
+bool GStreamerWebAudioPlayerClient::setEos()
+{
+    GST_DEBUG("entry:");
+
+    bool result = false;
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            if (mIsOpen)
+            {
+                if (mSampleDataBuffer.size())
+                {
+                    pushSamples();
+                }
+                m_pushSamplesTimer.cancel();
+                result = mClientBackend->setEos();
+            }
+            else
+            {
+                GST_ERROR("No web audio backend");
+            }
+        });
+
+    return result;
+}
+
+void GStreamerWebAudioPlayerClient::notifyPushSamplesTimerExpired()
+{
+    mBackendQueue.callInEventLoop([&]() { pushSamples(); });
+}
+
+bool GStreamerWebAudioPlayerClient::notifyNewSample()
+{
+    GST_DEBUG("entry:");
+
+    mBackendQueue.callInEventLoop(
+        [&]()
+        {
+            m_pushSamplesTimer.cancel();
+            std::vector<uint8_t> bufferData = getNextBufferData();
+            mSampleDataBuffer.insert(mSampleDataBuffer.end(), bufferData.begin(), bufferData.end());
+            pushSamples();
+        });
+
+    return true;
+}
+
+void GStreamerWebAudioPlayerClient::pushSamples()
+{
+    GST_DEBUG("entry:");
+    if (!mIsOpen || mSampleDataBuffer.empty())
+    {
+        return;
+    }
+
+    uint32_t availableFrames = 0u;
+    if (mClientBackend->getBufferAvailable(availableFrames))
+    {
+        auto dataToPush = std::min(availableFrames * 4, mSampleDataBuffer.size());
+        if ((dataToPush >= 0))
+        {
+            if (mClientBackend->writeBuffer(dataToPush / 4, mSampleDataBuffer.data()))
+            {
+                // remove pushed data from mSampleDataBuffer
+                if (dataToPush < mSampleDataBuffer.size())
+                {
+                    // to compact the memory copy everything to a new vector and swap.
+                    std::vector<uint8_t>(mSampleDataBuffer.begin() + dataToPush, mSampleDataBuffer.end())
```

Comment:
It might have quite big impact on performance in some bad scenarios, right?
Wouldn't it be more efficient to use std::queue or mSampleDataBuffer and create new temporary vector, which would be passed to writeBuffer?
Or change API of writeBuffer to pass readonly std::queue?